hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala

```diff
@@ -428,7 +429,9 @@ object ServiceBackendSocketAPI2 {
     val input = argv(5)
     val output = argv(6)
 
-    val backend = new ServiceBackend(revision, jarLocation, name, scratchDir)
+    // FIXME: when can the classloader be shared? (optimizer benefits!)
```

Comment:
Is this because of class-naming issues? I assume anything "client"-side (i.e. anything that's not batch code) should be able to share as much as it wants from a security perspective.

Reply:
I removed the FIXME. Explaining this above has made pretty clear to me when it can be shared. There is a hierarchy of security perimeters:
1. Global
2. Billing Project
3. User
4. Batch
5. Job

And a tree of sets of code. Each code set relies on code from one level higher. The levels of the tree are:
1. Java stdlibs
2. Spark JARs
2. Hail JAR for a given SHA
3. Hail generated code for a given SHA and batch and job

The question is: in which perimeter does each code-set live? In main the answer is:
1. Java stdlibs: Global
2. Spark JARs: Global
3. Hail JAR: Job
4. Hail generated code: Job

In this PR the answer is:
1. Java stdlibs: Global
2. Spark JARs: Global
3. Hail JAR: Global
4. Hail generated code: Job

The higher up the security hierarchy a code-set lives, the higher the chance that someone else has already paid the cost of JIT'ing a class you need. Descending the security hierarchy obviously gives fewer chances for bad code to screw with you.

Honestly, I think everything should be in the global hierarchy, because all this code is either written by Hail team or generated by code written by Hail team. However, we need to fix the naming issue with generated classes before code-set 4 can be shared across different batches.

Reply:
@johnc1231 I think this framing was really instructive for me, you might read over it.

Reply:
When you say everything should be in the global hierarchy, I think that's not quite true, since there's still the issue of different hail shas. The Java stdlib stuff should be sharable globally, but that's not true of two different hail SHAs with different versions of some random SType class.

Reply:
You're definitely right John. I think my picture is not quite right.